### PR TITLE
feat: anchored pref optimization

### PR DESCRIPTION
**Pull request by karel-contextual**
Add APO objectives, specifically equation 7 and 8 of the APO paper (https://huggingface.co/papers/2408.06266)

### CODE REVIEW

**Comment by qgallouedec** [comment]
#1929

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
@karel-contextual  can you kindly run:

```
pre-commit run --all-files
```

in the root dir of TRL to fix up the formatting

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1191,9 +1191,34 @@ def dpo_loss(
...
+
+            losses_chosen = F.sigmoid(self.beta * chosen_logratios) # Decrease chosen likelihood
+            losses_rejected = 1 - F.sigmoid(self.beta * (chosen_logratios - rejected_logratios)) # Decrease rejected likelihood more
+
+            losses = (
+                losses_chosen +
+                losses_rejected
+            )
+
+
```
Comment (kashif) [inline trl/trainer/dpo_trainer.py:1218]:
```suggestion
        elif self.loss_type == "apo_zero":
            # Eqn (7) of the APO paper (https://huggingface.co/papers/2408.06266)
            # Use this loss when you believe the chosen outputs are better than your model's default output

            losses_chosen = 1 - F.sigmoid(self.beta * chosen_logratios)  # Increase chosen likelihood
            losses_rejected = F.sigmoid(self.beta * rejected_logratios)  # Decrease rejected likelihood

            losses = losses_chosen + losses_rejected

        elif self.loss_type == "apo_down":
            # Eqn (8) of the APO paper (https://huggingface.co/papers/2408.06266)
            # Use this loss when you believe the chosen outputs are worse than your model's default output

            losses_chosen = F.sigmoid(self.beta * chosen_logratios)  # Decrease chosen likelihood
            losses_rejected = 1 - F.sigmoid(
                self.beta * (chosen_logratios - rejected_logratios)
            )  # Decrease rejected likelihood more

            losses = losses_chosen + losses_rejected

```

**Comment by qgallouedec** [comment]
LGTM now, thanks @karel-contextual!

**Comment by suhas1999** [comment]
@karel-contextual 

thanks for adding the APO code in the dpo_trainer.
In apo_down, I’d like to understand why losses_rejected is defined as:

1 - F.sigmoid(self.beta * (chosen_logratios - rejected_logratios))

instead of:

-F.sigmoid(self.beta * (chosen_logratios - rejected_logratios))?

I understand that the gradient doesn’t change, but I want to know why the 1 was added

**Pull request merged**